Limit the link job pool size based on the amount of RAM on Windows

hjyamauchi opened the pull request:
Mitigate the OOM thrashing and machine reboot issue during the
compiler build with debug info enabled on Windows.

Share the link job pool between LLVM and Swift unless overridden.

hjyamauchi commented:
We have been having machine hang issues during the compiler build on Windows when debug info (pdb) is enabled. This change limits the number parallel links based on the amount of RAM and I don't get the issue in my experiments.

compnerd commented on utils/build.ps1:
```diff
@@ -2099,6 +2099,17 @@ function Build-Runtime([Platform]$Platform, $Arch) {
+      $ParallelLinkJobs = [math]::Round((Get-WmiObject Win32_ComputerSystem).TotalPhysicalMemory / 16GB)
+      Write-Host "ParallelLinkJobs $ParallelLinkJobs"
+      if ((Get-WmiObject Win32_ComputerSystem).NumberOfLogicalProcessors -gt $ParallelLinkJobs) {
+        $DebugOptions = @{ LLVM_PARALLEL_LINK_JOBS = "$ParallelLinkJobs"; }
```
I'm not so sure about this. I've found limiting just the Swift linker pool is sufficient. Can you test with that?

hjyamauchi replied:
How did you limit the swift linker pool the way you did?

I tested with `LLVM_PARALLEL_LINK_JOBS=4  SWIFT_PARALLEL_LINK_JOBS=4` which didn't address it as it will result in 4 + 4 = 8 links running in parallel. I think sharing the link pool would be nicer because for example when you have many incremental link builds on the Swift side but none on the LLVM side, the Swift side would be limited to half of the pool and the other half would be not utilized.

Every bit might help and a simpler solution may be okay though :) Whatever works for you

hjyamauchi replied:
### Benchmarking

I benchmarked between the two schemes.

A Just passing`SWIFT_PARALLEL_LINK_JOBS=4` in the `Build-Compilers` job.
B  Sharing the LLVM-Swift link job pool and `LLVM_PARALLEL_LINK_JOBS=4` (as currently in this PR)

### Experiment steps
1. Do a full build with debug info enabled with `-DebugBuild`.
2. Touch `swift/lib/ClangImporter/ClangIncludePaths.cpp`.
3. Delete `S:/b/5/CMakeCaches.txt` and `S:/b/5/CMakeFiles/rules.ninja` so that the pool setting is deleted.
4. Apply the patch (for each approach)
5. Do an increment build and watch and measure the elapsed time for the `Build-Compilers` step.

This was run on a 64 GB Machine with 32 logical cores.

### Observations

#### A
- I observed 6 parallel links at max.
- I didn't see machine reboots but saw hang/unresponsiveness over Remote Desktop.
- This is likely better than the status quo where I usually observe 8 parallel links at max. 
- The link pool size is 32 (# cores) for LLVM and 4 for Swift and they aren't shared. 
- There seems to be additional limit enforced (by ninja?) that led to 8 instead of 32 parallel links. 
- In particular, the linking of `lld.exe` seemed to take a long time.
<img width="771" alt="Screenshot 2025-03-07 at 12 36 36 PM" src="https://github.com/user-attachments/assets/93114238-8c35-449b-9345-6c953372cb49" />
<img width="738" alt="Screenshot 2025-03-07 at 12 36 23 PM" src="https://github.com/user-attachments/assets/492f24b9-bfea-4606-9cf7-ddb242a22ec7" />


#### B
- I observed 4 parallel links at max.
- I didn't see reboots or hang/unresponsiveness over Remote Desktop.
- The link pool size is 4 for LLVM/Swift, shared. 
- The linking of `lld.exe` seemed to be faster.
<img width="756" alt="Screenshot 2025-03-07 at 12 14 06 PM" src="https://github.com/user-attachments/assets/a8ed9927-c6ed-4209-92de-0353311854b4" />
<img width="748" alt="Screenshot 2025-03-07 at 12 13 58 PM" src="https://github.com/user-attachments/assets/b99c8b8d-6ffa-4bbe-be33-a3659475bf3f" />


### The elapsed times of the `Build-Compilers` step

#### A
```
Run #1 11m2s
Run #2 10m13s
```

#### B
```
Run #1 2m6s
Run #2 3m42s
```

This seems to suggest a significant performance difference caused by the amount of the memory thrashing (swapping).

hjyamauchi commented:
Oh sorry this was mis-patched when I created a separate branch for this PR. It should add DebugOptions to the compiler build, not the swift runtime build. Will fix.

hjyamauchi committed:
Limit the link job pool size based on the amount of RAM on Windows

Mitigate the OOM thrashing and machine reboot issue during the
compiler build with debug info enabled on Windows.

Share the link job pool between LLVM and Swift unless overridden.

hjyamauchi commented:
> Oh sorry this was mis-patched when I created a separate branch for this PR. It should add DebugOptions to the compiler build, not the swift runtime build. Will fix.

This was fixed
